WebAPI: Add API description from Controller's POD

alvarocarvajald opened the pull request:
This PR adds the package OpenQA::WebAPI::Description which contains methods to be used on lib/OpenQA/WebAPI.pm to parse the Controller/API/V1 module files for POD documentation, and extract from there API descriptions to be shown in the GUI's help page.

In order to parse POD, the module Pod::POM was also added to the required modules for the project.

POD documentation was added to three API controller modules:

lib/OpenQA/WebAPI/Controller/API/V1/Feature.pm
lib/OpenQA/WebAPI/Controller/API/V1/JobGroup.pm
lib/OpenQA/WebAPI/Controller/API/V1/Worker.pm
lib/OpenQA/WebAPI/Controller/API/V1/Job.pm
lib/OpenQA/WebAPI/Controller/API/V1/Locks.pm

Adding POD documentation to the remaining modules, should populate API descriptions in the help page automatically.

okurz commented:
wow, that looks very promising. I would like to test that.

okurz commented:
https://build.opensuse.org/request/show/563765 created to provide the dependency for openSUSE

foursixnine commented on lib/OpenQA/WebAPI.pm:
```diff
@@ -553,5 +571,110 @@ sub run {
     Mojolicious::Commands->start_app('OpenQA::WebAPI');
 }
 
+sub get_pod_from_controllers {
```
I'd suggest to move this to it's own package along with _get_pod_text so that webAPI has a single responsability

alvarocarvajald replied:
So ... OpenQA::WebAPI::Documentation?

foursixnine commented on lib/OpenQA/WebAPI.pm:
```diff
@@ -553,5 +571,110 @@ sub run {
+                $desc = _get_pod_text($i) if $i->is_ordinary();
+            }
+        }
+        $methods_description{$controller . '#' . $methodname} = $desc;
```
I'd suggest to build the srting first, and then use it to avoid perl critic complains: https://travis-ci.org/os-autoinst/openQA/jobs/327749143#L652

alvarocarvajald replied:
Cool. Will do that.

alvarocarvajald commented:
@foursixnine Decided on OpenQA::WebAPI::Description. Let me know if this package name and location is OK.

foursixnine commented on lib/OpenQA/WebAPI/Description.pm:
```diff
@@ -0,0 +1,155 @@
+# Copyright (C) 2012-2017 SUSE LLC
```
2018?

coolo commented:
don't merge this before we have a package for pod::tree

alvarocarvajald commented:
Will investigate Pod::POM as an alternative to Pod::Tree. perl-Pod-POM already exists as a package for opensuse, so if the functionality is equivalent, should be a better option.

alvarocarvajald commented:
@coolo, @foursixnine, @okurz: reimplemented the POD walking methods in `OpenQA::WebAPI::Description` to use `Pod::POM` instead of `Pod::Tree`. `Pod::POM` has a package in opensuse, so it should be better to handle dependencies with this one.

foursixnine commented:
@alvarocarvajald Awesome! :) Remember to tidy up too :)

alvarocarvajald commented:
@foursixnine I did tidy up. :)

From what I can see in my system, files that are complaining on the tidy test are:

./lib/OpenQA/Resource/Jobs.pm
./lib/OpenQA/Schema/Result/Jobs.pm
./lib/OpenQA/WebAPI/Auth/OpenID.pm
./lib/OpenQA/WebAPI/Controller/Admin/JobGroup.pm
./lib/OpenQA/WebAPI/Controller/Admin/User.pm
./lib/OpenQA/WebAPI/Controller/Main.pm
./lib/OpenQA/WebAPI/Controller/Running.pm
./lib/OpenQA/WebAPI/Controller/Test.pm
./lib/OpenQA/WebAPI/Plugin/Gru.pm
./lib/OpenQA/Worker/Commands.pm
./lib/OpenQA/Worker.pm

Which aren't touched by this PR. Should I tidy those as well as part of this PR?

okurz commented:
@coolo did a tidy commit recently. Maybe you just need to rebase?

coolo commented:
I guess he uses an old perltidy

alvarocarvajald commented:
Updated perltidy as suggested and fixed a small issue in lib/OpenQA/WebAPI/Controller/API/V1/Locks.pm. Now it's not reporting any files in my local system ... let's see what reports travis.

coolo commented:
I don't mind merging it - but I consider 34 commits a bit tough - please squash to 1-5 commits

alvarocarvajald commented:
Squashed to 8 to keep some order between the different commits. Hope that's ok.

coolo commented:
ok, fair compromise :)
